xknx/io/gateway_scanner.py

```diff
@@ -172,6 +172,8 @@ def _response_rec_callback(self, knx_ip_frame: KNXIPFrame, udp_client: UDPClient
     def _add_found_gateway(self, gateway):
         if self.scan_filter.match(gateway):
             self.found_gateways.append(gateway)
+            if self.stop_on_found is None or self.stop_on_found <= 0:
```

Comment:
Isnt the `self.stop_on_found <= 0` a parameter validation which should not happen within this function? (I would place it within the initializer and throw an exception if the value is negative).

It also took my a while to understand the logic and would suggest to combine this into one condition? What do you think?

Reply:
Indeed. That would certainly be better (the existing implementation also never checked the actual value).
However, since this is a public field, it can still legally be changed after the instance has been constructed, so the validation cannot only be done in the constructor.
Even during scanning the value can theoretically be changed (although that is probably not very likely).
What is the general attitude to thinks like this in the project? I wouldn't immediately say it's the user's problem.

Assuming the value does not change after the `scan` method is called, a different way of solving this problem may be just keeping a protected field with a 'clean' upper limit value. Let's say that clean value is stored in `self._count_upper_bound` and is always a valid non-negative integer, then 0 indicates no limit and the rest indicates that there is one.
Then the condition I added to `_add_found_gateway` can be moved to the `scan` method, only to clean the current value, and the condition that compares the length can be changed to `if 0 < self._count_upper_bound <= len(self.found_gateways):`.

To be honest, this still does not feel like the best solution. More refactoring may be necessary.
Alternatively, changing the `stop_on_found` field to be protected would solve this problem immediately (and I don't think API compatibility is an issue yet).